File: Language_Essentials/python/q5_sort_comparators.py
# Custom Sorting and Comparators - TCS NQT
from functools import cmp_to_key

def cmp_descending(a, b):
    if a > b: return -1 # Less than zero means a < b in sort order
    if a < b: return 1  # Greater than zero means a > b in sort order
    return 0

def sort_ops():
    lst = [15, 2, 8, 22, 10]
    
    # 1. Default Sort (Ascending)
    lst.sort()
    print("Ascending:", lst)
    
    # 2. Sort Descending
    lst.sort(reverse=True)
    print("Descending (reverse=True):", lst)
    
    # 3. Sort using custom key
    # Key function takes an element and returns a value to sort by
    lst_strs = ["apple", "Banana", "cherry", "Date"]
    lst_strs.sort(key=str.lower)
    print("Sorted ignoring case:", lst_strs)
    
    # 3b. Sort with custom cmp_to_key
    lst_nums = [15, 2, 8, 22, 10]
    lst_nums.sort(key=cmp_to_key(cmp_descending))
    print("Descending (cmp_to_key):", lst_nums)
    
    # 4. Sorting lists of tuples/pairs
    pairs = [(1, 50), (2, 10), (3, 30)]
    
    # Sort pairs by second element ascending
    pairs.sort(key=lambda p: p[1])
    
    print("Pairs sorted by second value:", pairs)

if __name__ == "__main__":
    sort_ops()

File: Language_Essentials/java/Q5SortComparators.java
// Custom Sorting and Comparators - TCS NQT
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;

public class Q5SortComparators {
    static class Pair {
        int first, second;
        Pair(int f, int s) { first = f; second = s; }
    }

    public static void main(String[] args) {
        ArrayList<Integer> v = new ArrayList<>();
        Collections.addAll(v, 15, 2, 8, 22, 10);
        
        // 1. Default Sort (Ascending)
        Collections.sort(v);
        System.out.println("Ascending: " + v);
        
        // 2. Sort Descending using Collections.reverseOrder()
        Collections.sort(v, Collections.reverseOrder());
        System.out.println("Descending: " + v);
        
        // 3. Sorting Array of Primitives (ascending only built-in)
        int[] arr = {15, 2, 8, 22, 10};
        Arrays.sort(arr);
        System.out.println("Ascending array: " + Arrays.toString(arr));
        
        // Note: To sort primitives descending, you must use Integer[] or manually reverse
        
        // 4. Sort Descending using custom comparator
        ArrayList<Integer> v2 = new ArrayList<>(v);
        v2.sort((a, b) -> Integer.compare(b, a));
        System.out.println("Descending (custom lambda): " + v2);
        
        // 5. Sorting objects (pairs)
        ArrayList<Pair> pairs = new ArrayList<>();
        pairs.add(new Pair(1, 50));
        pairs.add(new Pair(2, 10));
        pairs.add(new Pair(3, 30));
        
        // Sort pairs by second value ascending
        pairs.sort(Comparator.comparingInt(p -> p.second));
        
        System.out.print("Pairs sorted by second value: ");
        for(Pair p : pairs) {
            System.out.print("{" + p.first + ", " + p.second + "} ");
        }
        System.out.println();
    }
}
